### PR TITLE
feat: allow async onBeforeUploadBegin callback

### PR DESCRIPTION
**Pull request by ibrahimyaacob92**
- PR is created to counter the issue faced when trying to validate the uploaded files with async function

image

### CODE REVIEW

File: packages/react/src/useUploadThing.ts [hunk 1]
```diff
@@ -14,6 +14,7 @@ import type {
   inferEndpointOutput,
   inferErrorShape,
 } from "uploadthing/server";
+import type { MaybePromise } from "../../uploadthing/src/internal/types";
```
Comment (juliusmarminge) [inline packages/react/src/useUploadThing.ts:17]: This import path wont work. I dont know if we export it from a public entrypoint, if not you can just inline the union type i guess

**Review by juliusmarminge — CHANGES_REQUESTED** [review]

File: packages/react/src/useUploadThing.ts (hunk 1 above)
Comment (ibrahimyaacob92) [reply packages/react/src/useUploadThing.ts]: @juliusmarminge updated

**Review by juliusmarminge — APPROVED** [review]

File: packages/react/src/useUploadThing.ts [hunk 2]
```diff
@@ -42,7 +42,7 @@ export type UseUploadthingProps<
   onUploadProgress?: (p: number) => void;
   onUploadError?: (e: UploadThingError<inferErrorShape<TRouter>>) => void;
   onUploadBegin?: (fileName: string) => void;
-  onBeforeUploadBegin?: (files: File[]) => File[];
+  onBeforeUploadBegin?:  (files: File[]) => Promise<File[]> | File;
```
Comment (juliusmarminge) [inline packages/react/src/useUploadThing.ts:45]: File[]

Comment (juliusmarminge) [reply packages/react/src/useUploadThing.ts:45]:
```suggestion
  onBeforeUploadBegin?: (files: File[]) => Promise<File[]> | File[];
```

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
Just formatting left, i dont have my laptop at hand so cant do it myself

**Review by juliusmarminge — APPROVED** [review]

**Pull request merged**